test/v3ext.c

```diff
@@ -307,6 +307,7 @@ static int test_addr_fam_len(void)
     testresult = 1;
   end:
     sk_IPAddressFamily_pop_free(addr, IPAddressFamily_free);
+    IPAddressFamily_free(f1);
```

Comment:
I think theres a risk of a double free in this case.  If in the following code:
```
if ((f1 = IPAddressFamily_new()) == NULL)
        goto end;   <==== (1)
    if (f1->ipAddressChoice == NULL &&
        (f1->ipAddressChoice = IPAddressChoice_new()) == NULL)
        goto end;   <==== (2)
    if (f1->addressFamily == NULL &&
        (f1->addressFamily = ASN1_OCTET_STRING_new()) == NULL)
        goto end;   <==== (3)
    if (!ASN1_OCTET_STRING_set(f1->addressFamily, key, keylen))
        goto end; <==== (4)
    if (!sk_IPAddressFamily_push(addr, f1))
        goto end;   <==== (5)

/* Shouldn't be able to canonize this as the len is > 3*/
    if (!TEST_false(X509v3_addr_canonize(addr)))
        goto end; <===(6)

```
A failure at locations 1-5  will be ok, but a failure at location 6 will result in 2 frees (one from the call tosk_IPAddressFamily_pop_free, as f1 has been pushed onto the stack struct prior to the goto at (6), and then again to free f1 directly.

I think once you push the element to the stack , you likely want to set f1 to null to avoid that

Reply:
This sort of fragility is why I was tempted to introduce safe-by-construction macros to correctly transfer ownership of the memory from a detached pointer to the stack. Yes, overkill in context, but plausibly the right sort of *habit*. https://github.com/openssl/openssl/pull/23991#issuecomment-2024480415